### PR TITLE
How to connect two networked Zephyr instances together

### PR DESCRIPTION
**Pull request by jukkar**
Document how to connect multiple QEMU or native_posix Zephyr instances together.

### CODE REVIEW

**Comment by pfalcon** [comment]
We already have https://docs.zephyrproject.org/latest/guides/networking/qemu_setup.html#running-multiple-qemu-vms-of-the-same-sample . Does the new content added mashes well with that section?

**Comment by jukkar** [comment]
> Does the new content added mashes well with that section?

This PR generalizes the multi-instance setup, the old instructions are mainly for echo-server/client connectivity with very specific setup using SLIP.

File: drivers/ethernet/Kconfig [hunk 1]
```diff
@@ -32,6 +32,15 @@ config ETH_QEMU_IFACE_NAME
 	  script from net-tools project can be used to create the network
 	  interface.
 
+config ETH_QEMU_EXTRA_ARGS
```
Comment (bwasim) [inline drivers/ethernet/Kconfig:35]: I think we should consider merging this and `NET_QEMU_USER_EXTRA_ARGS` for a consolidated Kconfig option.. Both are used to convey similar type of information in different scenarios.. Thoughts ?

Comment (jukkar) [reply drivers/ethernet/Kconfig:35]: Well, atm they are mutually exclusive (cannot exists at the same time) so is no big issue here. Also you might want to have different options for them so switch between user/tap mode would be cumbersome.

File: cmake/app/boilerplate.cmake [hunk 2]
```diff
@@ -556,8 +556,13 @@ if(CONFIG_QEMU_TARGET)
       Enable Qemu supported ethernet driver like e1000 at drivers/ethernet"
     )
   elseif(CONFIG_NET_QEMU_ETHERNET)
+    if(CONFIG_ETH_QEMU_EXTRA_ARGS)
+      set(NET_QEMU_ETH_EXTRA_ARGS ",${CONFIG_ETH_QEMU_EXTRA_ARGS}")
+    else()
+      set(NET_QEMU_ETH_EXTRA_ARGS "")
```
Comment (tejlmand) [inline cmake/app/boilerplate.cmake:562]: Those two lines can be deleted.
If `${NET_QEMU_ETH_EXTRA_ARGS}` is not defined below when used, then it's simply empty.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: drivers/ethernet/Kconfig (hunk 1 above)
Comment (tejlmand) [reply drivers/ethernet/Kconfig:35]: Regarding the extra arguments, and the help text below, then there is no way for the user to know if he should provide a `;`-separated (CMake style),  `,`-separated, or `' '`-separated list in this field.
From the CMake code, I can see the only valid separator to use when providing multiple args is `,`.

Would it be possible to syntax validate the extra args, before passing it to QEMU ?

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: drivers/ethernet/Kconfig (hunk 1 above)
Comment (jukkar) [reply drivers/ethernet/Kconfig:35]: > From the CMake code, I can see the only valid separator to use when providing multiple args is `,`.
> 
> Would it be possible to syntax validate the extra args, before passing it to QEMU ?

I can add more text to the help string. I think we can leave syntax validation to qemu, seems pointless to do that in both in cmake file and qemu, when the qemu does that anyway.

**Comment by jukkar** [comment]
Updated according to comments.

**Review by pfalcon — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
LGTM, thank you.

File: drivers/ethernet/Kconfig [hunk 3]
```diff
@@ -32,6 +32,18 @@ config ETH_QEMU_IFACE_NAME
...
+	string "Extra arguments to QEMU -nic option"
+	depends on NET_QEMU_ETHERNET
+	default ""
+	help
+	  Extra arguments passed to QEMU -nic option when Ethernet Networking
+	  is enabled. Typically this is used to set the network MAC address of
+	  Zephyr instance. This option can contain multiple QEMU option
+	  arguments. Each QEMU argument must be separated by comma ","
```
Comment (tejlmand) [inline drivers/ethernet/Kconfig:43]:
```suggestion
 	  arguments. Each QEMU argument must be separated by comma "," and no spaces
```

Have experienced people before thinking that they can and spaces in such cases, and write:
`arg0, arg1` which will clearly break in this situation.

**Review by tejlmand — APPROVED** [review]
Look good.
A minor comment, I would like to see addressed, but nothing blocking.

**Pull request merged**